node/src/actors/chain_manager/handlers.rs

```diff
@@ -126,129 +115,144 @@ impl Handler<EpochNotification<EveryEpochPayload>> for ChainManager {
                 }
             }
             StateMachine::Synchronizing => {}
-            StateMachine::Synced => match self.chain_state {
-                ChainState {
-                    chain_info: Some(ref mut chain_info),
-                    reputation_engine: Some(ref mut rep_engine),
-                    ..
-                } => {
-                    if self.epoch_constants.is_none()
-                        || self.vrf_ctx.is_none()
-                        || self.secp.is_none()
-                    {
-                        log::error!("{}", ChainManagerError::ChainNotReady);
-                        return;
-                    }
-                    // Decide the best candidate
-                    let target_vrf_slots = VrfSlots::from_rf(
-                        u32::try_from(rep_engine.ars().active_identities_number()).unwrap(),
-                        chain_info.consensus_constants.mining_replication_factor,
-                        chain_info.consensus_constants.mining_backup_factor,
-                    );
-                    // TODO: replace for loop with a try_fold
-                    let mut chosen_candidate = None;
-                    for (key, (block_candidate, vrf_proof)) in self
-                        .candidates
-                        .drain()
-                        .sorted_by_key(|(_key, (_block_candidate, vrf_proof))| *vrf_proof)
-                    {
-                        let block_pkh = &block_candidate.block_sig.public_key.pkh();
-                        let reputation = rep_engine.trs().get(block_pkh);
-
-                        if let Some((chosen_key, chosen_reputation, chosen_vrf_proof, _, _)) =
-                            chosen_candidate
+            StateMachine::Synced | StateMachine::Ready => {
+                if self.sm_state == StateMachine::Synced && current_epoch > 0 {
+                    // We only reach Ready state after genesis event
```

Comment:
What's the issue with being in Ready state in epoch 0?

Reply:
that you will be able to mine, and the epoch 0 is for the genesis block

Reply:
Then we can move this check to `mining.rs`, as

    if self.sm_state != StateMachine::Ready || current_epoch == 0